File: extensions/database/src/com/google/refine/extension/database/DatabaseImportController.java
/*
 * Copyright (c) 2017, Tony Opara
 *        All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without 
 * modification, are permitted provided that the following conditions are met:
 * - Redistributions of source code must retain the above copyright notice, this 
 *   list of conditions and the following disclaimer.
 * - Redistributions in binary form must reproduce the above copyright notice, 
 *   this list of conditions and the following disclaimer in the documentation 
 *   and/or other materials provided with the distribution.
 * 
 * Neither the name of Google nor the names of its contributors may be used to 
 * endorse or promote products derived from this software without specific 
 * prior written permission.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" 
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, 
 * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR 
 * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR 
 * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, 
 * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, 
 * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; 
 * OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, 
 * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR 
 * OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF 
 * ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

package com.google.refine.extension.database;

import java.io.IOException;
import java.io.Writer;
import java.util.LinkedList;
import java.util.List;
import java.util.Properties;

import javax.servlet.ServletException;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.fasterxml.jackson.core.JsonGenerator;
import com.fasterxml.jackson.databind.node.ObjectNode;
import com.google.refine.ProjectManager;
import com.google.refine.ProjectMetadata;
import com.google.refine.RefineServlet;
import com.google.refine.commands.HttpUtilities;
import com.google.refine.extension.database.model.DatabaseColumn;
import com.google.refine.extension.database.model.DatabaseQueryInfo;
import com.google.refine.importers.TabularImportingParserBase;
import com.google.refine.importing.ImportingController;
import com.google.refine.importing.ImportingJob;
import com.google.refine.importing.ImportingManager;
import com.google.refine.model.Project;
import com.google.refine.util.JSONUtilities;
import com.google.refine.util.ParsingUtilities;

public class DatabaseImportController implements ImportingController {

    private static final Logger logger = LoggerFactory.getLogger("DatabaseImportController");
    protected RefineServlet servlet;
    public static int DEFAULT_PREVIEW_LIMIT = 100;
    public static String OPTIONS_KEY = "options";

    @Override
    public void init(RefineServlet servlet) {
        this.servlet = servlet;
    }

    @Override
    public void doGet(HttpServletRequest request, HttpServletResponse response)
            throws ServletException, IOException {
        HttpUtilities.respond(response, "error", "GET not implemented");
    }

    @Override
    public void doPost(HttpServletRequest request, HttpServletResponse response)
            throws ServletException, IOException {
        if (logger.isDebugEnabled()) {
            logger.debug("doPost Query String::{}", request.getQueryString());
        }
        response.setCharacterEncoding("UTF-8");
        Properties parameters = ParsingUtilities.parseUrlParameters(request);

        String subCommand = parameters.getProperty("subCommand");

        if (logger.isDebugEnabled()) {
            logger.info("doPost::subCommand::{}", subCommand);
        }

        if ("initialize-parser-ui".equals(subCommand)) {
            doInitializeParserUI(request, response, parameters);
        } else if ("parse-preview".equals(subCommand)) {
            try {

                doParsePreview(request, response, parameters);

            } catch (DatabaseServiceException e) {
                logger.error("doPost::DatabaseServiceException::{}", e);
                HttpUtilities.respond(response, "error", getDbServiceException(e));
            }
        } else if ("create-project".equals(subCommand)) {
            doCreateProject(request, response, parameters);
        } else {
            HttpUtilities.respond(response, "error", "No such sub command");
        }

    }

    private String getDbServiceException(Exception ex) {
        String message = "";
        if (ex instanceof DatabaseServiceException) {
            DatabaseServiceException dbEx = (DatabaseServiceException) ex;
            if (dbEx.isSqlException()) {
                message = message + dbEx.getSqlCode() + " " + dbEx.getSqlState();
            }
        }
        message = message + ex.getMessage();

        return message;
    }

    /**
     * 
     * @param request
     * @param response
     * @param parameters
     * @throws ServletException
     * @throws IOException
     */
    private void doInitializeParserUI(HttpServletRequest request, HttpServletResponse response, Properties parameters)
            throws ServletException, IOException {
        if (logger.isDebugEnabled()) {
            logger.debug("::doInitializeParserUI::");
        }

        ObjectNode result = ParsingUtilities.mapper.createObjectNode();
        ObjectNode options = ParsingUtilities.mapper.createObjectNode();
        JSONUtilities.safePut(result, "status", "ok");
        JSONUtilities.safePut(result, OPTIONS_KEY, options);

        JSONUtilities.safePut(options, "skipDataLines", 0);
        JSONUtilities.safePut(options, "storeBlankRows", true);
        JSONUtilities.safePut(options, "storeBlankCellsAsNulls", true);
        if (logger.isDebugEnabled()) {
            logger.debug("doInitializeParserUI:::{}", result.toString());
        }

        HttpUtilities.respond(response, result.toString());

    }

    /**
     * doParsePreview
     * 
     * @param request
     * @param response
     * @param parameters
     * @throws ServletException
     * @throws IOException
     * @throws DatabaseServiceException
     */
    private void doParsePreview(
            HttpServletRequest request, HttpServletResponse response, Properties parameters)
            throws ServletException, IOException, DatabaseServiceException {
        if (logger.isDebugEnabled()) {
            logger.debug("JobID::{}", parameters.getProperty("jobID"));
        }

        long jobID = Long.parseLong(parameters.getProperty("jobID"));
        ImportingJob job = ImportingManager.getJob(jobID);
        if (job == null) {
            HttpUtilities.respond(response, "error", "No such import job");
            return;
        }

        DatabaseQueryInfo databaseQueryInfo = getQueryInfo(request);

        if (databaseQueryInfo == null) {
            HttpUtilities.respond(response, "error", "Invalid or missing Query Info");
        }

        job.updating = true;
        try {
            ObjectNode optionObj = ParsingUtilities.evaluateJsonStringToObjectNode(
                    request.getParameter("options"));

            List<Exception> exceptions = new LinkedList<Exception>();

            job.prepareNewProject();

            parsePreview(
                    databaseQueryInfo,
                    job.project,
                    job.metadata,
                    job,
                    DEFAULT_PREVIEW_LIMIT,
                    optionObj,
                    exceptions);
            Writer w = response.getWriter();
            JsonGenerator writer = ParsingUtilities.mapper.getFactory().createGenerator(w);
            try {
                writer.writeStartObject();
                if (exceptions.size() == 0) {
                    job.project.update(); // update all internal models, indexes, caches, etc.
                    writer.writeStringField("status", "ok");
                } else {
                    writer.writeStringField("status", "error");
                    writer.writeStringField("message", getExceptionString(exceptions));
                }
                writer.writeEndObject();
            } catch (IOException e) {
                throw new ServletException(e);
            } finally {
                writer.flush();
                writer.close();
                w.flush();
                w.close();
            }

        } catch (IOException e) {
            throw new ServletException(e);
        } finally {
            job.touch();
            job.updating = false;
        }
    }

    private String getExceptionString(List<Exception> exceptions) {
        String ex = "";
        for (Exception e : exceptions) {
            ex = ex + e.getLocalizedMessage() + "\n";
        }
        // TODO Auto-generated method stub
        return ex;
    }

    /**
     * 
     * @param dbQueryInfo
     * @param project
     * @param metadata
     * @param job
     * @param limit
     * @param options
     * @param exceptions
     * @throws DatabaseServiceException
     */
    private static void parsePreview(
            DatabaseQueryInfo dbQueryInfo,
            Project project,
            ProjectMetadata metadata,
            final ImportingJob job,
            int limit,
            ObjectNode options,
            List<Exception> exceptions) throws DatabaseServiceException {

        DatabaseService databaseService = DatabaseService.get(dbQueryInfo.getDbConfig().getDatabaseType());
        String querySource = getQuerySource(dbQueryInfo);

        List<DatabaseColumn> columns = databaseService.getColumns(dbQueryInfo.getDbConfig(), dbQueryInfo.getQuery());

        setProgress(job, querySource, -1);

        JSONUtilities.safePut(options, "ignoreLines", 0); // number of blank lines at the beginning to ignore
        JSONUtilities.safePut(options, "headerLines", 1); // number of header lines

        TabularImportingParserBase.readTable(
                project,
                metadata,
                job,
                new DBQueryResultPreviewReader(job, databaseService, querySource, columns, dbQueryInfo, 100),
                querySource,
                limit,
                options,
                exceptions);

        setProgress(job, querySource, 100);

    }

    /**
     * doCreateProject
     * 
     * @param request
     * @param response
     * @param parameters
     */
    private void doCreateProject(HttpServletRequest request, HttpServletResponse response, Properties parameters)
            throws ServletException, IOException {
        if (logger.isDebugEnabled()) {
            logger.debug("DatabaseImportController::doCreateProject:::{}", parameters.getProperty("jobID"));
        }

        long jobID = Long.parseLong(parameters.getProperty("jobID"));
        final ImportingJob job = ImportingManager.getJob(jobID);
        if (job == null) {
            HttpUtilities.respond(response, "error", "No such import job");
            return;
        }

        final DatabaseQueryInfo databaseQueryInfo = getQueryInfo(request);
        if (databaseQueryInfo == null) {
            HttpUtilities.respond(response, "error", "Invalid or missing Query Info");
        }

        job.updating = true;
        try {
            final ObjectNode optionObj = ParsingUtilities.evaluateJsonStringToObjectNode(
                    request.getParameter("options"));

            final List<Exception> exceptions = new LinkedList<Exception>();

            job.setState("creating-project");

            final Project project = new Project();

            new Thread() {

                @Override
                public void run() {
                    ProjectMetadata pm = new ProjectMetadata();
                    pm.setName(JSONUtilities.getString(optionObj, "projectName", "Untitled"));
                    pm.setEncoding(JSONUtilities.getString(optionObj, "encoding", "UTF-8"));

                    try {
                        parseCreate(
                                databaseQueryInfo,
                                project,
                                pm,
                                job,
                                -1,
                                optionObj,
                                exceptions);
                    } catch (DatabaseServiceException e) {
                        logger.error("DatabaseImportController::doCreateProject:::run{}", e);
                        // throw new RuntimeException("DatabaseServiceException::", e);
                    }

                    if (!job.canceled) {
                        if (exceptions.size() > 0) {
                            job.setError(exceptions);
                        } else {
                            project.update(); // update all internal models, indexes, caches, etc.
                            ProjectManager.singleton.registerProject(project, pm);
                            job.setState("created-project");
                            job.setProjectID(project.id);
                            // logger.info("DatabaseImportController::doCreateProject:::run::projectID :{}",
                            // project.id);
                        }

                        job.touch();
                        job.updating = false;
                    }
                }
            }.start();

            HttpUtilities.respond(response, "ok", "done");
        } catch (IOException e) {
            throw new ServletException(e);
        }
    }

    /**
     * @param dbQueryInfo
     * @param project
     * @param metadata
     * @param job
     * @param limit
     * @param options
     * @param exceptions
     * @throws DatabaseServiceException
     */
    private static void parseCreate(
            DatabaseQueryInfo dbQueryInfo,
            Project project,
            ProjectMetadata metadata,
            final ImportingJob job,
            int limit,
            ObjectNode options,
            List<Exception> exceptions) throws DatabaseServiceException {

        DatabaseService databaseService = DatabaseService.get(dbQueryInfo.getDbConfig().getDatabaseType());
        String querySource = getQuerySource(dbQueryInfo);

        List<DatabaseColumn> columns = databaseService.getColumns(dbQueryInfo.getDbConfig(), dbQueryInfo.getQuery());

        setProgress(job, querySource, -1);

        JSONUtilities.safePut(options, "ignoreLines", 0); // number of blank lines at the beginning to ignore
        JSONUtilities.safePut(options, "headerLines", 1); // number of header lines

        long startTime = System.currentTimeMillis();

        TabularImportingParserBase.readTable(
                project,
                metadata,
                job,
                new DBQueryResultImportReader(job, databaseService, querySource, columns, dbQueryInfo, getCreateBatchSize()),
                querySource,
                limit,
                options,
                exceptions);

        long endTime = System.currentTimeMillis();
        if (logger.isDebugEnabled()) {
            logger.debug("Execution Time: {}", endTime - startTime);
        }

        setProgress(job, querySource, 100);

    }

    private static int getCreateBatchSize() {
        String propBatchSize = DatabaseModuleImpl.getImportCreateBatchSize();
        int batchSize = 100;
        if (propBatchSize != null && !propBatchSize.isEmpty()) {
            try {
                batchSize = Integer.parseInt(propBatchSize);
            } catch (NumberFormatException nfe) {

            }
        }
        return batchSize;
    }

    /**
     * @param request
     * @return
     */
    private DatabaseQueryInfo getQueryInfo(HttpServletRequest request) {
        DatabaseConfiguration jdbcConfig = new DatabaseConfiguration();
        jdbcConfig.setConnectionName(request.getParameter("connectionName"));
        jdbcConfig.setDatabaseType(request.getParameter("databaseType"));
        jdbcConfig.setDatabaseHost(request.getParameter("databaseServer"));
        try {
            jdbcConfig.setDatabasePort(Integer.parseInt(request.getParameter("databasePort")));
        } catch (NumberFormatException nfE) {
            logger.error("getQueryInfo :: invalid database port ::{}", nfE);
        }
        jdbcConfig.setDatabaseUser(request.getParameter("databaseUser"));
        jdbcConfig.setDatabasePassword(request.getParameter("databasePassword"));
        jdbcConfig.setDatabaseName(request.getParameter("initialDatabase"));
        jdbcConfig.setDatabaseSchema(request.getParameter("initialSchema"));

        String query = request.getParameter("query");
        if (logger.isDebugEnabled()) {
            logger.debug("jdbcConfig::{}, query::{}", jdbcConfig, query);
        }
        if (jdbcConfig.getDatabaseHost() == null || jdbcConfig.getDatabaseName() == null
                || jdbcConfig.getDatabasePassword() == null || jdbcConfig.getDatabaseType() == null
                || jdbcConfig.getDatabaseUser() == null || query == null
                || (jdbcConfig.getDatabasePort() == 0 && !"SQLite".equalsIgnoreCase(jdbcConfig.getDatabaseType()))) {
            if (logger.isDebugEnabled()) {
                logger.debug("Missing Database Configuration::{}", jdbcConfig);
            }
            return null;
        }

        return new DatabaseQueryInfo(jdbcConfig, query);
    }

    private static String getQuerySource(DatabaseQueryInfo dbQueryInfo) {
        String dbType = dbQueryInfo.getDbConfig().getDatabaseType();
        return DatabaseService.get(dbType).getDatabaseUrl(dbQueryInfo.getDbConfig());
    }

    private static void setProgress(ImportingJob job, String querySource, int percent) {
        job.setProgress(percent, "Reading " + querySource);
    }
}

File: extensions/database/module/scripts/index/database-import-controller.js
/*
 * Copyright (c) 2017, Tony Opara
 *        All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 * - Redistributions of source code must retain the above copyright notice, this
 *   list of conditions and the following disclaimer.
 * - Redistributions in binary form must reproduce the above copyright notice,
 *   this list of conditions and the following disclaimer in the documentation
 *   and/or other materials provided with the distribution.
 *
 * Neither the name of Google nor the names of its contributors may be used to
 * endorse or promote products derived from this software without specific
 * prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
 * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
 * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
 * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
 * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
 * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
 * OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
 * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
 * OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
 * ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

I18NUtil.init("database");

Refine.DatabaseImportController = function(createProjectUI) {
  this._createProjectUI = createProjectUI;

  this._parsingPanel = createProjectUI.addCustomPanel();

  createProjectUI.addSourceSelectionUI({
    label: $.i18n('database-import/importer-name'),
    id: "database-source",
    ui: new Refine.DatabaseSourceUI(this)
  });

};
Refine.CreateProjectUI.controllers.push(Refine.DatabaseImportController);

Refine.DatabaseImportController.prototype.startImportingDocument = function(queryInfo) {
    var dismiss = DialogSystem.showBusy($.i18n('database-import/preparing'));
    //alert(queryInfo.query);
    var self = this;

    Refine.postCSRF(
      "command/core/create-importing-job",
      null,
      function(data) {
        Refine.wrapCSRF(function(token) {
            $.post(
            "command/core/importing-controller?" + $.param({
                "controller": "database/database-import-controller",
                "subCommand": "initialize-parser-ui",
                "csrf_token": token
            }),
            queryInfo,

            function(data2) {
                dismiss();

                if (data2.status == 'ok') {
                self._queryInfo = queryInfo;
                self._jobID = data.jobID;
                self._options = data2.options;

                self._showParsingPanel();

                } else {
                alert(data2.message);
                }
            },
            "json"
            );
        });
      },
      "json"
    );
};

Refine.DatabaseImportController.prototype.getOptions = function() {
   var options = {

   };

    var parseIntDefault = function(s, def) {
      try {
        var n = parseInt(s);
        if (!isNaN(n)) {
          return n;
        }
      } catch (e) {
        // Ignore
      }
      return def;
    };


    if (this._parsingPanelElmts.skipCheckbox[0].checked) {
      options.skipDataLines = parseIntDefault(this._parsingPanelElmts.skipInput[0].value, 0);
    } else {
      options.skipDataLines = 0;
    }
    if (this._parsingPanelElmts.limitCheckbox[0].checked) {
      options.limit = parseIntDefault(this._parsingPanelElmts.limitInput[0].value, -1);
    } else {
      options.limit = -1;
    }
    options.storeBlankRows = this._parsingPanelElmts.storeBlankRowsCheckbox[0].checked;
    options.storeBlankCellsAsNulls = this._parsingPanelElmts.storeBlankCellsAsNullsCheckbox[0].checked;

    options.disableAutoPreview = this._parsingPanelElmts.disableAutoPreviewCheckbox[0].checked;

    return options;
};

Refine.DatabaseImportController.prototype._showParsingPanel = function() {
    var self = this;

    this._parsingPanel.off().empty().html(
        DOM.loadHTML("database",'scripts/index/database-parsing-panel.html'));

    this._parsingPanelElmts = DOM.bind(this._parsingPanel);

    this._parsingPanelElmts.startOverButton.html($.i18n('database-parsing/start-over'));
    this._parsingPanelElmts.database_conf_pars.html($.i18n('database-parsing/conf-pars'));
    this._parsingPanelElmts.database_proj_name.html($.i18n('database-parsing/proj-name'));
    this._parsingPanelElmts.createProjectButton.html($.i18n('database-parsing/create-proj'));
    this._parsingPanelElmts.database_options.html($.i18n('database-parsing/option'));
    this._parsingPanelElmts.previewButton.html($.i18n('database-parsing/preview-button'));
    this._parsingPanelElmts.database_updating.html($.i18n('database-parsing/updating-preview'));
    this._parsingPanelElmts.database_discard_next.html($.i18n('database-parsing/discard-next'));
    this._parsingPanelElmts.database_discard.html($.i18n('database-parsing/discard'));
    this._parsingPanelElmts.database_limit_next.html($.i18n('database-parsing/limit-next'));
    this._parsingPanelElmts.database_limit.html($.i18n('database-parsing/limit'));
    this._parsingPanelElmts.database_store_row.html($.i18n('database-parsing/store-row'));
    this._parsingPanelElmts.database_store_cell.html($.i18n('database-parsing/store-cell'));
    this._parsingPanelElmts.database_disable_auto_preview.text($.i18n('database-parsing/disable-auto-preview'));

    if (this._parsingPanelResizer) {
      $(window).off('resize', this._parsingPanelResizer);
    }

    this._parsingPanelResizer = function() {
      var elmts = self._parsingPanelElmts;
      var width = self._parsingPanel.width();
      var height = self._parsingPanel.height();
      var headerHeight = elmts.wizardHeader.outerHeight(true);
      var controlPanelHeight = 250;

      elmts.dataPanel
      .css("left", "0px")
      .css("top", headerHeight + "px")
      .css("width", (width - DOM.getHPaddings(elmts.dataPanel)) + "px")
      .css("height", (height - headerHeight - controlPanelHeight - DOM.getVPaddings(elmts.dataPanel)) + "px");
      elmts.progressPanel
      .css("left", "0px")
      .css("top", headerHeight + "px")
      .css("width", (width - DOM.getHPaddings(elmts.progressPanel)) + "px")
      .css("height", (height - headerHeight - controlPanelHeight - DOM.getVPaddings(elmts.progressPanel)) + "px");

      elmts.controlPanel
      .css("left", "0px")
      .css("top", (height - controlPanelHeight) + "px")
      .css("width", (width - DOM.getHPaddings(elmts.controlPanel)) + "px")
      .css("height", (controlPanelHeight - DOM.getVPaddings(elmts.controlPanel)) + "px");
    };

    $(window).on('resize',this._parsingPanelResizer);
    this._parsingPanelResizer();

    this._parsingPanelElmts.startOverButton.on('click',function() {
      // explicitly cancel the import job
      Refine.CreateProjectUI.cancelImportingJob(self._jobID);

      delete self._jobID;
      delete self._options;

      self._createProjectUI.showSourceSelectionPanel();
    });

    this._parsingPanelElmts.createProjectButton.on('click',function() { self._createProject(); });
    this._parsingPanelElmts.previewButton.on('click',function() { self._updatePreview(); });
    //alert("datetime::" + $.now());
    //this._parsingPanelElmts.projectNameInput[0].value = this._queryInfo.connectionName + "_" + this._queryInfo.databaseUser + "_" + $.now();
    this._parsingPanelElmts.projectNameInput[0].value = this._queryInfo.databaseServer +  "_" + this._queryInfo.initialDatabase + "_" + $.now();


    if (this._options.limit > 0) {
      this._parsingPanelElmts.limitCheckbox.prop("checked", true);
      this._parsingPanelElmts.limitInput[0].value = this._options.limit.toString();
    }
    if (this._options.skipDataLines > 0) {
      this._parsingPanelElmts.skipCheckbox.prop("checked", true);
      this._parsingPanelElmts.skipInput.value[0].value = this._options.skipDataLines.toString();
    }
    if (this._options.storeBlankRows) {
      this._parsingPanelElmts.storeBlankRowsCheckbox.prop("checked", true);
    }
    if (this._options.storeBlankCellsAsNulls) {
      this._parsingPanelElmts.storeBlankCellsAsNullsCheckbox.prop("checked", true);
    }

    if (this._options.disableAutoPreview) {
      this._parsingPanelElmts.disableAutoPreviewCheckbox.prop('checked', true);
    }

    // If disableAutoPreviewCheckbox is not checked, we will schedule an automatic update
    var onChange = function() {
      if (!self._parsingPanelElmts.disableAutoPreviewCheckbox[0].checked)
      {
        self._scheduleUpdatePreview();
      }
    };
    this._parsingPanel.find("input").on("change", onChange);
    this._parsingPanel.find("select").on("change", onChange);

    this._createProjectUI.showCustomPanel(this._parsingPanel);
    this._updatePreview();
};

Refine.DatabaseImportController.prototype._scheduleUpdatePreview = function() {
    if (this._timerID != null) {
      window.clearTimeout(this._timerID);
      this._timerID = null;
    }

    var self = this;
    this._timerID = window.setTimeout(function() {
      self._timerID = null;
      self._updatePreview();
    }, 500); // 0.5 second
  };

Refine.DatabaseImportController.prototype._updatePreview = function() {
    var self = this;
   // alert("query::" + this._queryInfo.query);
    this._parsingPanelElmts.dataPanel.hide();
    this._parsingPanelElmts.progressPanel.show();
    this._queryInfo.options = JSON.stringify(this.getOptions());
    //alert("options:" + this._queryInfo.options);

    Refine.wrapCSRF(function(token) {
        $.post(
        "command/core/importing-controller?" + $.param({
            "controller": "database/database-import-controller",
            "jobID": self._jobID,
            "subCommand": "parse-preview",
            "csrf_token": token
        }),

        self._queryInfo,

        function(result) {
            if (result.status == "ok") {
                self._getPreviewData(function(projectData) {
                self._parsingPanelElmts.progressPanel.hide();
                self._parsingPanelElmts.dataPanel.show();

                new Refine.PreviewTable(projectData, self._parsingPanelElmts.dataPanel.off().empty());
            });
            } else {

            alert('Errors:\n' +  (result.message) ? result.message : Refine.CreateProjectUI.composeErrorMessage(job));
            self._parsingPanelElmts.progressPanel.hide();

            Refine.CreateProjectUI.cancelImportingJob(self._jobID);

            delete self._jobID;
            delete self._options;

            self._createProjectUI.showSourceSelectionPanel();


            }
        },
        "json"
        );
    });
  };

Refine.DatabaseImportController.prototype._getPreviewData = function(callback, numRows) {
    var self = this;
    var result = {};

    $.post(
      "command/core/get-models?" + $.param({ "importingJobID" : this._jobID }),
      null,
      function(data) {
        for (var n in data) {
          if (data.hasOwnProperty(n)) {
            result[n] = data[n];
          }
        }

        $.post(
          "command/core/get-rows?" + $.param({
            "importingJobID" : self._jobID,
            "start" : 0,
            "limit" : numRows || 100 // More than we parse for preview anyway
          }),
          null,
          function(data) {
            result.rowModel = data;
            callback(result);
          },
          "jsonp"
        );
      },
      "json"
    );
};

Refine.DatabaseImportController.prototype._createProject = function() {
    var projectName = jQueryTrim(this._parsingPanelElmts.projectNameInput[0].value);
    if (projectName.length == 0) {
      window.alert("Please name the project.");
      this._parsingPanelElmts.projectNameInput.focus();
      return;
    }

    var self = this;
    var options = this.getOptions();
    options.projectName = projectName;

    this._queryInfo.options = JSON.stringify(options);
    Refine.wrapCSRF(function(token) {
        $.post(
        "command/core/importing-controller?" + $.param({
            "controller": "database/database-import-controller",
            "jobID": self._jobID,
            "subCommand": "create-project",
            "csrf_token": token
        }),
        self._queryInfo,
        function(o) {
            if (o.status == 'error') {
            alert(o.message);
            } else {
            var start = new Date();
            var timerID = window.setInterval(
                function() {
                self._createProjectUI.pollImportJob(
                    start,
                    self._jobID,
                    timerID,
                    function(job) {
                        return "projectID" in job.config;
                    },
                    function(jobID, job) {
                        //alert("jobID::" + jobID + " job :" + job);
                        window.clearInterval(timerID);
                        Refine.CreateProjectUI.cancelImportingJob(jobID);
                        document.location = "project?project=" + job.config.projectID;
                    },
                    function(job) {
                        alert(Refine.CreateProjectUI.composeErrorMessage(job));
                    }
                );
                },
                1000
            );
            self._createProjectUI.showImportProgressPanel($.i18n('database-import/creating'), function() {
                // stop the timed polling
                window.clearInterval(timerID);

                // explicitly cancel the import job
                Refine.CreateProjectUI.cancelImportingJob(jobID);

                self._createProjectUI.showSourceSelectionPanel();
            });
            }
        },
        "json"
        );
    });
};
